mdpow/tests/test_automated_dihedral_analysis.py

```diff
@@ -0,0 +1,215 @@
+
+        # default relative tolerance is 1e-6 for pytest.approx
+        dh1_mean == pytest.approx(self.DG_O1C2N3S4_mean)
+        dh1_var == pytest.approx(self.DG_O1C2N3S4_var)
+
+        dh2_mean == pytest.approx(self.DG_C13141520_mean)
+        dh2_var == pytest.approx(self.DG_C13141520_var)
+        
+    def test_save_df(self, gen_data, SM25_tmp_dir):
+        dihedrals.save_df(df=gen_data[2], df_save_dir=SM25_tmp_dir, molname='SM25')
+        assert SM25_tmp_dir / 'SM25' / 'SM25_full_df.bz2'
+                                        
+
+    @pytest.mark.skipif(sys.version_info < (3, 8), reason="scipy circvar gives wrong answers") 
+    def test_periodic_angle(self, gen_data):
+
+        df_aug = gen_data[3]
+
+        aug_dh2_result = df_aug.loc[df_aug['selection'] == 'C13-C14-C15-C20']['dihedral']
+
+        aug_dh2_mean = circmean(aug_dh2_result, high=180, low=-180)
+        aug_dh2_var = circvar(aug_dh2_result, high=180, low=-180)
+
+        aug_dh2_mean == pytest.approx(self.ADG_C13141520_mean)
+        aug_dh2_var == pytest.approx(self.ADG_C13141520_var)
+
+    def test_save_fig(self, SM25_tmp_dir):
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=SM25_tmp_dir,
+                                              resname=self.resname, molname='SM25',
+                                              solvents=('water',))
+        assert SM25_tmp_dir / 'SM25' / 'SM25_C10-C5-S4-O11_violins.pdf'
+
+    def test_DataFrame_input(self, SM25_tmp_dir):
+        test_df = pd.DataFrame([['C1-C2-C3-C4', 'water', 'Coulomb', 0, 0, 60.0],
+                                ['C1-C2-C3-C5', 'water', 'Coulomb', 0, 0, 60.0]],
+                                [1,2],['selection', 'solvent', 'interaction', 'lambda', 'time', 'dihedral'])
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=SM25_tmp_dir,
+                                              resname=self.resname,
+                                              solvents=('water',), dataframe=test_df)
+
+        with open('mdpow.log', 'r') as file:
+            log_content = file.read()
+            return_value = log_content.find('Proceeding with results DataFrame provided.')
+
+        assert return_value != -1
+
+    def test_figdir_warning(self, SM25_tmp_dir):
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, figdir=None,
+                                              resname=self.resname, molname='SM25',
+                                              solvents=('water',))
+
+        with open('mdpow.log', 'r') as file:
+            log_content = file.read()
+            return_value = log_content.find('Figures will not be saved unless figdir kwarg '
+                                            'is specified, otherwise, continue without saving.')
+
+        assert return_value != -1
+
+    def test_save_results_warning(self, SM25_tmp_dir):
+        dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, df_save_dir=None,
```

Comment:
same as above: instead of log file searching, have the code raise a warning and then use [pytest.warns](https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html#warns)
```python
with pytest.warns(UserWarning):
    xxx = dihedrals.automated_dihedral_analysis(dirname=SM25_tmp_dir, df_save_dir=None,

# assert something about xxx, too
```
You can also have `pytest.warns` check the warning message.

If you emit another warning than UserWarning, check for that instead.

Reply:
They are not actually warnings of the Warning type subclass of Exceptions, they are logging.WARNING and logging.INFO messages, that are output or returned to display informative messages to the user and to also be recorded in the logs in case someone forgot to specify a results data frame saving directory and wants to know why their results were not saved.

I changed the testing methods, and they now use pytest's fixture `caplog` because `pytest.warn` only works with objects in the Warning class.